### PR TITLE
add ip4 implementation netserver

### PR DESCRIPTION
**Pull request by ArsenArsen**
done so far:
 * `SOCK_RAW` receive
 * IP assignment
 * SOCK_RAW send

to do (non exhaustive list):
 * fragmentation support
 * dynamic driver loading and discovery (in short, `dlopen()` the right `.so`s) 
 * syn cookies

### CODE REVIEW

File: libarch/include/arch/bit.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace arch {
+enum class endian {
+    little = __ORDER_LITTLE_ENDIAN__,
+    big    = __ORDER_BIG_ENDIAN__,
+    native = __BYTE_ORDER__
+};
+
+static_assert(endian::native == endian::little || endian::native == endian::big,
+		"who uses middle endian?");
```
Comment (avdgrinten) [inline libarch/include/arch/bit.hpp:14]: Remove this message.

File: libarch/include/arch/bit.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	if constexpr (sizeof(T) == 2) {
+		return __builtin_bswap16(val);
+	} else if constexpr (sizeof(T) == 4) {
+		return __builtin_bswap32(val);
+	} else if constexpr (sizeof(T) == 8) {
+		return __builtin_bswap64(val);
+	}
+	static_assert("unsupported swap size");
+	// suppress warning, this code is unreachable
+	return val;
```
Comment (avdgrinten) [inline libarch/include/arch/bit.hpp:28]: `__builtin_unreachable()`

File: libarch/include/arch/bit.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+inline T bswap(T val) {
+	static_assert(std::is_integral_v<T>, "T must be an integral type");
+	if constexpr (sizeof(T) == 2) {
+		return __builtin_bswap16(val);
+	} else if constexpr (sizeof(T) == 4) {
+		return __builtin_bswap32(val);
+	} else if constexpr (sizeof(T) == 8) {
+		return __builtin_bswap64(val);
+	}
+	static_assert("unsupported swap size");
```
Comment (avdgrinten) [inline libarch/include/arch/bit.hpp:26]: Does this work? Doesn't this need some `dependent_false`.

File: meson.build [hunk 4]
```diff
@@ -66,6 +66,7 @@ if get_option('build_drivers')
 	subdir('core/drm/')
 	subdir('core/virtio/')
 	subdir('mbus/')
+	subdir('netserver/')
```
Comment (avdgrinten) [inline meson.build:69]: Put it into `drivers/` or add a new `servers/` directory.

File: netserver/netserver.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+
+name: nic-virtio
```
Comment (avdgrinten) [inline netserver/netserver.yml:2]: `netserver`

File: libarch/include/arch/dma_structs.hpp [hunk 6]
```diff
@@ -26,13 +27,16 @@ struct dma_buffer_view {
...
+
+	explicit dma_buffer_view(dma_pool *pool, std::byte *data, size_t size)
 	: _pool{pool}, _data{data}, _size{size} { }
 
 	size_t size() const {
 		return _size;
 	}
 
-	void *data() const {
+	std::byte *data() const {
```
Comment (avdgrinten) [inline libarch/include/arch/dma_structs.hpp:39]: This should remain `void *`. Byte-wise addressing might not always be what we want.

File: netserver/src/ip/ip4.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+#include <iomanip>
+#include <arch/bit.hpp>
+#include <protocols/fs/server.hpp>
+#include <queue>
+
+using namespace protocols::fs;
+
+namespace {
+template<typename T>
+bool takeData(T &field, arch::dma_buffer_view &container) {
```
Comment (avdgrinten) [inline netserver/src/ip/ip4.cpp:16]: This function has a very unspecific name.

File: netserver/src/ip/ip4.cpp [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+	uint8_t ttl;
+	uint8_t protocol;
+	uint16_t checksum;
+
+	uint32_t source;
+
+	uint32_t destination;
+
+	arch::dma_buffer_view capsule() {
```
Comment (avdgrinten) [inline netserver/src/ip/ip4.cpp:53]: Why `capsule()`? Would maybe `payload` be better?

File: netserver/src/ip/ip4.cpp [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+const FileOperations *sockops() {
+	static FileOperations ops = {
+		.recvMsg = [] (void *obj, const char *creds,
+			uint32_t flags, void *data, size_t len,
+			void *addr_buf, size_t addr_size, size_t max_ctrl_len) {
+			auto self = static_cast<Socket *>(obj);
+			std::cout << "netserver: recvmsg" << std::endl;
+			return self->recvmsg(creds, flags, data, len, addr_buf,
+				addr_size, max_ctrl_len);
+		}
```
Comment (avdgrinten) [inline netserver/src/ip/ip4.cpp:161]: Unnecessary boilerplate. It's simpler to make the `recvMsg` function of the socket a static method and do the `static_cast` there.

File: netserver/src/main.cpp [hunk 10]
```diff
@@ -91,22 +79,39 @@ async::detached serve(helix::UniqueLane lane) {
...
 		auto conversation = accept.descriptor();
+		auto sendError = [&conversation] (managarm::fs::Errors err) -> async::result<void> {
+			helix::SendBuffer send_resp;
+
+			managarm::fs::SvrResponse resp;
+			resp.set_error(err);
+			auto ser = resp.SerializeAsString();
+			auto &&transmit = helix::submitAsync(conversation, helix::Dispatcher::global(),
+					helix::action(&send_resp, ser.data(), ser.size()));
+			co_await transmit.async_wait();
```
Comment (avdgrinten) [inline netserver/src/main.cpp:90]: This can use `helix_ng`.

**Comment by avdgrinten** [comment]
Also, private members need to be distinguished (probably through an `_` suffix).

File: libarch/include/arch/dma_structs.hpp [hunk 11]
```diff
@@ -26,13 +27,20 @@ struct dma_buffer_view {
...
 
 	size_t size() const {
 		return _size;
 	}
 
 	void *data() const {
+		return static_cast<void *>(_data);
+	}
+
+	std::byte *byteData() {
```
Comment (avdgrinten) [inline libarch/include/arch/dma_structs.hpp:43]: `byte_data` (libarch uses `snake_case`).

File: libarch/include/arch/dma_structs.hpp [hunk 12]
```diff
@@ -26,13 +27,20 @@ struct dma_buffer_view {
 	: _pool{nullptr}, _data{nullptr}, _size{0} { }
 
 	explicit dma_buffer_view(dma_pool *pool, void *data, size_t size)
+	: dma_buffer_view{pool, reinterpret_cast<std::byte *>(data), size} {}
+
+	explicit dma_buffer_view(dma_pool *pool, std::byte *data, size_t size)
```
Comment (avdgrinten) [inline libarch/include/arch/dma_structs.hpp:32]: This ctor is subsumed in the other one.

File: libarch/include/arch/dma_structs.hpp [hunk 13]
```diff
@@ -49,7 +57,7 @@ struct dma_buffer_view {
 
 private:
 	dma_pool *_pool;
-	void *_data;
+	std::byte *_data;
```
Comment (avdgrinten) [inline libarch/include/arch/dma_structs.hpp:60]: Revert this.

File: netserver/src/nic.cpp [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+async::detached runDevice(std::shared_ptr<nic::Link> dev) {
+	using namespace arch;
+	while(true) {
+		dma_buffer frameBuffer { dev->dmaPool(), 1514 };
+		co_await dev->receive(frameBuffer);
+		auto capsule = frameBuffer.subview(14);
+		auto data = reinterpret_cast<uint8_t*>(frameBuffer.data());
+		uint16_t ethertype = data[12] << 8 | data[13];
+		auto dest = reinterpret_cast<nic::MACAddress*>(data)[0];
+		auto src = reinterpret_cast<nic::MACAddress*>(data)[1];
```
Comment (avdgrinten) [inline netserver/src/nic.cpp:29]: `memcpy`.

File: netserver/src/nic.hpp [hunk 15]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <arch/dma_pool.hpp>
+#include <async/result.hpp>
+
+namespace nic {
+struct MACAddress {
```
Comment (avdgrinten) [inline netserver/src/nic.hpp:7]: `MacAddress`.

File: meson.build [hunk 16]
```diff
@@ -91,6 +90,10 @@ if get_option('build_drivers')
 	subdir('testsuites/posix-torture/')
 	subdir('testsuites/posix-tests/')
 
+	netserver_inc = include_directories('servers/netserver/include/')
```
Comment (avdgrinten) [inline meson.build:93]: Put this into `netserver`'s meson.build.

File: drivers/nic/virtio/src/virtio.cpp [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+	virtual async::result<void> receive(arch::dma_buffer_view) override;
+	virtual async::result<void> send(const arch::dma_buffer_view) override;
+	virtual arch::dma_pool *dmaPool() override;
+
+	virtual ~VirtioNic() override = default;
+private:
+	std::unique_ptr<virtio_core::Transport> m_transport;
+	arch::contiguous_pool m_dmaPool;
+	virtio_core::Queue *m_receiveVq;
+	virtio_core::Queue *m_transmitVq;
```
Comment (avdgrinten) [inline drivers/nic/virtio/src/virtio.cpp:49]: Update the coding convention here.

File: servers/netserver/src/ip/ip4.hpp [hunk 18]
```diff
@@ -4,19 +4,71 @@
...
+
+		inline Route &setGateway(uint32_t gateway) {
+			this->gateway = gateway;
+			return *this;
+		}
+
+		inline Route &setMetric(unsigned int metric) {
+			this->metric = metric;
+			return *this;
+		}
```
Comment (avdgrinten) [inline servers/netserver/src/ip/ip4.hpp:35]: Remote this (non-idiomatic boilerplate).

File: protocols/fs/fs.proto [hunk 19]
```diff
@@ -123,7 +123,7 @@ message CntRequest {
 	// used by SOCKET_CREATE
 	optional int64 type = 49;
 	optional int64 protocol = 50;
-	optional int64 domain = 54;
+	optional int64 domain = 59;
```
Comment (avdgrinten) [inline protocols/fs/fs.proto]: Why was this changed?

File: protocols/fs/src/server.cpp [hunk 20]
```diff
@@ -419,10 +419,11 @@ async::detached handlePassthrough(smarter::shared_ptr<void> file,
...
 
-		auto result = co_await file_ops->recvMsg(file.get(), extract_creds.credentials(), req.flags(),
-				buffer.data(), buffer.size(),
-				addr.data(), addr.size(),
-				req.ctrl_size());
+		auto result = co_await file_ops->recvMsg(file.get(),
+			extract_creds.credentials(), req.flags(),
+			buffer.data(), buffer.size(),
+			addr.data(), addr.size(),
+			req.ctrl_size());
```
Comment (avdgrinten) [inline protocols/fs/src/server.cpp:426]: No change here?

File: protocols/fs/fs.proto (hunk 19 above)
Comment (ArsenArsen) [reply protocols/fs/fs.proto]: conflict with commits on master when I rebased (inode_source in CntRequest)

File: protocols/fs/src/server.cpp (hunk 20 above)
Comment (ArsenArsen) [reply protocols/fs/src/server.cpp:426]: reformatted the line to make it shorter and removed an extraneous indentation level

**Comment by avdgrinten** [comment]
Commit 6dcb15eecac92149b697714affcc7351f28a1b08 (`netserver: make dmaPool() nonvirtual`) deletes two unrelated files (?).

File: servers/netserver/src/ip/arp.cpp [hunk 21]
```diff
@@ -0,0 +1,199 @@
+#include "arp.hpp"
+
+#include <helix/ipc.hpp>
+#include <helix/timer.hpp>
+#include <arch/bit.hpp>
+#include <iomanip>
+#include "ip4.hpp"
+
+struct ArpLeader {
```
Comment (avdgrinten) [inline servers/netserver/src/ip/arp.cpp:9]: Rename to `ArpHeader`.

File: servers/netserver/src/ip/arp.cpp [hunk 22]
```diff
@@ -154,6 +154,10 @@ void Neighbours::updateTable(uint32_t ip, nic::MacAddress mac) {
 
 namespace {
 async::detached entryProber(uint32_t ip, Neighbours::Entry &e, uint32_t sender) {
+	if (e.state == Neighbours::State::probe) {
+		// probing already, no need to run
+		co_return;
+	}
```
Comment (avdgrinten) [inline servers/netserver/src/ip/arp.cpp:160]: It would be better to not even enter this method (to avoid allocating the coroutine frame).

File: servers/netserver/include/netserver/nic.hpp [hunk 23]
```diff
@@ -45,6 +49,8 @@ struct Link {
 	//! Sends an entire ethernet frame
 	virtual async::result<void> send(const arch::dma_buffer_view) = 0;
 	arch::dma_pool *dmaPool();
+	virtual AllocatedBuffer allocateFrame(MacAddress to, EtherType type,
+		size_t payloadSize);
```
Comment (avdgrinten) [inline servers/netserver/include/netserver/nic.hpp:53]: Let's make this non-virtual until we have real users of this functionality.

File: servers/netserver/src/ip/ip4.cpp [hunk 24]
```diff
@@ -119,13 +121,32 @@ class Ip4Packet {
...
 	explicit Ip4Socket(int proto) : proto(proto) {}
 
 	static async::result<RecvResult> recvmsg(void *obj,
 			const char *creds,
 			uint32_t flags, void *data, size_t len,
 			void *addr_buf, size_t addr_size, size_t max_ctrl_len) {
+		auto hton = [] (auto x) -> decltype(x) {
+			using namespace arch;
+			return convert_endian<endian::big>(x);
+		};
```
Comment (avdgrinten) [inline servers/netserver/src/ip/ip4.cpp:149]: Instead of adding boilerplate here, use the arch:: function directly.

File: meson.build (hunk 16 above)
Comment (avdgrinten) [reply meson.build:93]: This is still not fixed.

File: drivers/nic/virtio/src/virtio.cpp (hunk 17 above)
Comment (avdgrinten) [reply drivers/nic/virtio/src/virtio.cpp:48]: This is still not fixed.

File: protocols/fs/src/server.cpp [hunk 25]
```diff
@@ -440,11 +441,12 @@ async::detached handlePassthrough(smarter::shared_ptr<void> file,
 		}
 
 		auto data = std::get<RecvData>(result);
+		resp.set_addr_size(data.addressLength);
 		auto ser = resp.SerializeAsString();
 		auto [send_resp, send_addr, send_data, send_ctrl]
 				= co_await helix_ng::exchangeMsgs(conversation,
 			helix_ng::sendBuffer(ser.data(), ser.size()),
-			helix_ng::sendBuffer(addr.data(), data.addressLength),
+			helix_ng::sendBuffer(addr.data(), addr.size()),
```
Comment (avdgrinten) [inline protocols/fs/src/server.cpp:449]: This should be `std::min(addr.size(), data.addressLength)`.

File: servers/netserver/src/ip/arp.hpp [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+		reachable,
+		stale
+	};
+	struct Entry {
+		uint64_t mtime_ns;
+		nic::MacAddress mac;
+		async::doorbell change;
+		State state = State::none;
+	};
+	async::result<Entry*> tryResolve(uint32_t addr, uint32_t sender);
```
Comment (avdgrinten) [inline servers/netserver/src/ip/arp.hpp:22]: This still returns an `Entry *`.

File: servers/netserver/src/ip/arp.hpp [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+	Entry &getEntry(uint32_t addr);
```
Comment (avdgrinten) [inline servers/netserver/src/ip/arp.hpp:23]: This should probably be private.
